File: usecase-2.1/work/pipeline.jl
#!/usr/bin/env julia

# This pipeline computes averages from the participants.tsv file
#
# Use as 
#    ./pipeline.jl [options] <inputdir> <outputdir> <level>
# where the input and output directory must be specified, and the 
# level is either "group" or "participant".
#
# Optional arguments:
#   -h,--help           Show this help and exit.
#   --verbose           Enable verbose output.
#   --start-idx <num>   Start index for participant selection.
#   --stop-idx <num>    Stop index for participant selection.

# This code is shared under the CC0 license
#
# Copyright (C) 2024, SIESTA workpackage 15 team

using ArgParse
using CSV
using DataFrames
using Statistics

############################################################
# Parse the command line arguments
############################################################

function parse_commandline()
    s = ArgParseSettings()

    @add_arg_table s begin
        "inputdir"
            help = "the input directory"
            required = true
        "outputdir"
            help = "the output directory"
            required = true
        "level"
            help = "the analysis level, either 'participant' or 'group'"
            required = true
        "--start-idx"
            help = "index of the first participant to include, one-offset"
            arg_type = Int
            default = 0
        "--stop-idx"
            help = "index of the last participant to include, one-offset"
            arg_type = Int
            default = 0
        "--verbose", "-v"
            help = "give more verbose information for debugging"
            action = :store_true
    end

    return parse_args(s)
end

##########################################################################
# Compute the averages of the age, height, and weight of the participants
##########################################################################

function main(options)

    if haskey(options, "verbose") && options["verbose"]
        println("options =")
        display(options)
    end

    if haskey(options, "level") && options["level"] == "participant"
        println("nothing to do at the participant level")
        return
    end

    # Create the output directory and its parents if they don't exist
    mkpath(options["outputdir"])

    inputfile  = joinpath(options["inputdir"], "participants.tsv")
    outputfile = joinpath(options["outputdir"], "results.tsv")

    # Read the participants.tsv file into a DataFrame
    participants = CSV.read(inputfile, DataFrame; delim='\t', missingstring="n/a")

    if haskey(options, "verbose") && options["verbose"]
        println("data contains $(nrow(participants)) participants")
    end

    # Select participants based on start_idx and stop_idx
    if haskey(options, "stop_idx") && options["stop_idx"] > 0
        participants = participants[1:options["stop_idx"], :]
    end
    if haskey(options, "start_idx") && options["start_idx"] > 0
        participants = participants[options["start_idx"]:end, :]
    end

    if haskey(options, "verbose") && options["verbose"]
        println("selected $(nrow(participants)) participants")
    end

    # Compute averages
    averagedAge    = mean(skipmissing(participants.age))
    averagedHeight = mean(skipmissing(participants.Height))
    averagedWeight = mean(skipmissing(participants.Weight))

    # Put the results in a DataFrame
    result = DataFrame(
        averagedAge = [averagedAge],
        averagedHeight = [averagedHeight],
        averagedWeight = [averagedWeight]
    )

    if haskey(options, "verbose") && options["verbose"]
        display(result)
    end

    # Write the results to a TSV file
    CSV.write(outputfile, result; delim='\t', writeheader=false)
end

##########################################################################
# execute the code if it is run as a script
##########################################################################

options = parse_commandline()
main(options)

File: usecase-2.1/work/pipeline.py
#!/usr/bin/env python3

"""This pipeline computes averages from the participants.tsv file

This code is shared under the CC0 license

Copyright (C) 2024, SIESTA workpackage 15 team
"""

import pandas as pd
import argparse
from pathlib import Path


##########################################################################
# Compute the averages of the age, height, and weight of the participants
##########################################################################

def main(options: dict):
    """
    This function computes averages from the participants.tsv file

    The Options keys:
    -----------------
      - inputdir:   Directory containing participants.tsv (str)
      - outputdir:  Directory to save results.tsv (str)
      - level:      Analysis level, either "group" or "participant" (str)
      - verbose:    Enable verbose output (bool)
      - start-idx:  Start index for participant selection (int)
      - stop-idx:   Stop index for participant selection (int)
    """

    if options.get('verbose'):
        print('options =')
        print(options)

    if options.get('level') == 'participant':
        print("nothing to do at the participant level")
        return

    # Create the output directory and its parents if they don't exist
    Path(options['outputdir']).mkdir(parents=True, exist_ok=True)

    # Read the participants.tsv input file into a DataFrame
    inputfile  = Path(options['inputdir'])/'participants.tsv'
    outputfile = Path(options['outputdir'])/'results.tsv'
    if not inputfile.is_file():
        print(f"WARNING: input file does not exist: {inputfile}")
        return
    participants = pd.read_csv(inputfile, sep='\t')
    if options.get('verbose'):
        print(f"data contains {len(participants)} participants")

    # Select participants based on start_idx and stop_idx, these are specified using 1-indexing
    print(options.get('start_idx'))
    print(options.get('stop_idx'))
    if options.get('stop_idx') is not None:
        participants = participants.iloc[:(options['stop_idx'])]
    if options.get('start_idx') is not None:
        participants = participants.iloc[(options['start_idx']-1):]
    if options.get('verbose'):
        print(f"selected {len(participants)} participants")

    # Compute averages
    averaged_age    = participants['age'].mean(skipna=True)
    averaged_height = participants['Height'].mean(skipna=True)
    averaged_weight = participants['Weight'].mean(skipna=True)

    # Put the results in a DataFrame
    result = pd.DataFrame({
        'averagedAge': [averaged_age],
        'averagedHeight': [averaged_height],
        'averagedWeight': [averaged_weight]
    })
    if options.get('verbose'):
        print(result)

    # Write the results to a TSV file
    result.to_csv(outputfile, sep='\t', index=False, header=False)


##########################################################################
# execute the code if it is run as a script
##########################################################################

if __name__ == "__main__":

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('inputdir', type=str, help='Directory containing participants.tsv')
    parser.add_argument('outputdir', type=str, help='Directory to save results.tsv')
    parser.add_argument('level', type=str, help='The analysis level', choices=['participant', 'group'])
    parser.add_argument('-v', '--verbose', action='store_true', help='Enable verbose output')
    parser.add_argument('--start-idx', type=int, default=None, help='Start index for participant selection')
    parser.add_argument('--stop-idx', type=int, default=None, help='Stop index for participant selection')

    main(vars(parser.parse_args()))
